File: LeetCode/Python Solutions/Smallest String Starting From Leaf/smallest.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def smallestFromLeaf(self, root: TreeNode) -> str:
        self.smallest = "~"  # Use a character lexicographically higher than 'z'

        def dfs(node, path):
            if node:
                # Prepend current character to path
                path = chr(node.val + ord('a')) + path
                # If it's a leaf node
                if not node.left and not node.right:
                    if path < self.smallest:
                        self.smallest = path
                else:
                    dfs(node.left, path)
                    dfs(node.right, path)

        dfs(root, "")
        return self.smallest

File: LeetCode/Java Solutions/Smallest String Starting From Leaf/smallest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private String smallest = "~";  // Use a character lexicographically higher than 'z'

    public String smallestFromLeaf(TreeNode root) {
        dfs(root, "");
        return smallest;
    }

    private void dfs(TreeNode node, String path) {
        if (node != null) {
            path = (char)(node.val + 'a') + path;  // Prepend character
            if (node.left == null && node.right == null) {  // Leaf node
                if (path.compareTo(smallest) < 0) {
                    smallest = path;
                }
            } else {
                dfs(node.left, path);
                dfs(node.right, path);
            }
        }
    }
}